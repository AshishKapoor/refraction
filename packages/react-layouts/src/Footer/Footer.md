Footer component example:

```js
<Footer style={{ marginBottom: 10 }}>
  <Footer.Slot left>
    {paddingClass => <Button className={paddingClass}>How Quid Works</Button>}
  </Footer.Slot>
  <Footer.Slot center separated>
    {paddingClass => (
      <React.Fragment>
        <Button className={paddingClass}>Preview Slides</Button>
        <Button className={paddingClass} importance="primary">
          Share
        </Button>
      </React.Fragment>
    )}
  </Footer.Slot>
  <Footer.Slot right>
    {paddingClass => (
      <Button className={paddingClass}>Export all to PPT</Button>
    )}
  </Footer.Slot>
</Footer>
```

`Footer.Slot` component helps you to divide `Footer` into smaller pieces. It accepts `center`, `separated`, `left` and `right` props.
Use `paddingClass` argument to add extra padding to the first and last component inside `Footer.Slot`.

For transparent `Footer` use `transparent` prop.

```js
<Footer transparent style={{ marginBottom: 10 }}>
  <Footer.Slot left>
    {paddingClass => <Button className={paddingClass}>How Quid Works</Button>}
  </Footer.Slot>
  <Footer.Slot center separated>
    {paddingClass => (
      <React.Fragment>
        <Button className={paddingClass}>Preview Slides</Button>
        <Button className={paddingClass} importance="primary">
          Share
        </Button>
      </React.Fragment>
    )}
  </Footer.Slot>
  <Footer.Slot right>
    {paddingClass => (
      <Button className={paddingClass}>Export all to PPT</Button>
    )}
  </Footer.Slot>
</Footer>
```
