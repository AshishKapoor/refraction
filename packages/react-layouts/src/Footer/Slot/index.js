/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import { withFallback as wf, sizes } from '@quid/theme';
import css from '@emotion/css/macro';
import { ClassNames } from '@emotion/core';

const PADDING = sizes.small;

export type Props = {
  separated?: boolean,
  left?: boolean,
  right?: boolean,
  center?: boolean,
  children: string => React.Node,
};

const Slot = styled(
  ({
    separated = false,
    left = false,
    right = false,
    center = false,
    children,
    ...props
  }: Props) => (
    <ClassNames>
      {({
        css,
        paddingClass = css`
          margin-left: ${PADDING};
          margin-right: ${PADDING};
          &:first-of-type {
            margin-left: calc(${PADDING} * 2);
          }

          &:last-of-type {
            margin-right: calc(${PADDING} * 2);
          }
        `,
      }) => <div {...props}>{children(paddingClass)}</div>}
    </ClassNames>
  )
)`
  display: flex;
  align-items: center;


  ${props =>
    props.separated &&
    css`
      border-color: ${wf(props =>
        props.theme.current === 'light'
          ? props.theme.colors.gray1
          : props.theme.colors.gray5
      )(props)};
      border-style: solid;
      border-left-width: 1px;
      border-right-width: 1px;
      border-top-width: 0;
      border-bottom-width: 0;
      padding-left: ${PADDING};
      padding-right: ${PADDING};
    `}

  ${props =>
    props.left &&
    css`
      flex: 1;
      justify-content: flex-start;
    `}

  ${props =>
    props.right &&
    css`
      flex: 1;
      justify-content: flex-end;
    `}


  ${props =>
    props.center &&
    css`
      justify-content: center;
    `}
`;

export default Slot;
