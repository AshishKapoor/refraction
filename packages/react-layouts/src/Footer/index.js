/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import React from 'react';
import styled from '@emotion/styled/macro';
import { withFallback as wf } from '@quid/theme';
import css from '@emotion/css/macro';
import Slot from './Slot';

export type Props = {
  transparent?: boolean,
};

const Footer = styled(({ transparent, ...props }: Props) => <nav {...props} />)`
  font-size: 1rem;
  display: flex;
  height: 2.86em;
  align-items: stretch;

  ${wf(
    props =>
      !props.transparent &&
      props.theme.current === 'light' &&
      css`
        background-color: ${props.theme.colors.white};
      `
  )};

  ${wf(
    props =>
      !props.transparent &&
      props.theme.current === 'dark' &&
      css`
        background-color: ${props.theme.colors.gray7};
      `
  )};
`;

Footer.Slot = Slot;

export default Footer;
