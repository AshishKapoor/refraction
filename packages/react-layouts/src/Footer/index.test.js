/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import React, { Fragment } from 'react';
import { mount } from 'enzyme';
import Footer from './index';

it('renders the expected markup', () => {
  const wrapper = mount(<Footer>Lorem ipsum</Footer>);

  expect(wrapper).toMatchSnapshot();
});

it('renders the expected markup, transparent Footer', () => {
  const wrapper = mount(
    <Footer theme={{ current: 'dark', colors: { gray7: 'gray7' } }}>
      Lorem ipsum
    </Footer>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders the expected markup basic Footer.Slot', () => {
  const wrapper = mount(
    <Footer.Slot>
      {() => {
        'Test';
      }}
    </Footer.Slot>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders a left Footer.Slot', () => {
  const wrapper = mount(
    <Footer.Slot left>
      {() => {
        'Test';
      }}
    </Footer.Slot>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders a right Footer.Slot', () => {
  const wrapper = mount(
    <Footer.Slot right>
      {() => {
        'Test';
      }}
    </Footer.Slot>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders a center Footer.Slot', () => {
  const wrapper = mount(
    <Footer.Slot center>
      {() => {
        'Test';
      }}
    </Footer.Slot>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders a separated Footer.Slot', () => {
  const wrapper = mount(
    <Footer.Slot separated>
      {() => {
        'Test';
      }}
    </Footer.Slot>
  );

  expect(wrapper).toMatchSnapshot();
});

it('renders a separated Footer.Slot, dark theme', () => {
  const wrapper = mount(
    <Footer.Slot
      separated
      theme={{ current: 'dark', colors: { gray7: 'gray7' } }}
    >
      {() => {
        'Test';
      }}
    </Footer.Slot>
  );

  expect(wrapper).toMatchSnapshot();
});

it('should apply extra padding to first and last Slot child elements', () => {
  const wrapper = mount(
    <Footer.Slot>
      {paddingClass => (
        <Fragment>
          <button className={paddingClass}>Button 1</button>
          <button className={paddingClass}>Button 2</button>
          <button className={paddingClass}>Button 3</button>
          <button className={paddingClass}>Button 4</button>
          <button className={paddingClass}>Button 5</button>
        </Fragment>
      )}
    </Footer.Slot>
  );

  expect(wrapper).toMatchSnapshot();
});
