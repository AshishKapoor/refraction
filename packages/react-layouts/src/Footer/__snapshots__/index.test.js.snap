// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a center Footer.Slot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<Slot
  center={true}
>
  <ForwardRef(render)>
    <div
      className="emotion-0 emotion-1"
    />
  </ForwardRef(render)>
</Slot>
`;

exports[`renders a left Footer.Slot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

<Slot
  left={true}
>
  <ForwardRef(render)>
    <div
      className="emotion-0 emotion-1"
    />
  </ForwardRef(render)>
</Slot>
`;

exports[`renders a right Footer.Slot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

<Slot
  right={true}
>
  <ForwardRef(render)>
    <div
      className="emotion-0 emotion-1"
    />
  </ForwardRef(render)>
</Slot>
`;

exports[`renders a separated Footer.Slot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-color: #E3E6E8;
  border-style: solid;
  border-left-width: 1px;
  border-right-width: 1px;
  border-top-width: 0;
  border-bottom-width: 0;
  padding-left: 5px;
  padding-right: 5px;
}

<Slot
  separated={true}
>
  <ForwardRef(render)>
    <div
      className="emotion-0 emotion-1"
    />
  </ForwardRef(render)>
</Slot>
`;

exports[`renders a separated Footer.Slot, dark theme 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-style: solid;
  border-left-width: 1px;
  border-right-width: 1px;
  border-top-width: 0;
  border-bottom-width: 0;
  padding-left: 5px;
  padding-right: 5px;
}

<Slot
  separated={true}
  theme={
    Object {
      "colors": Object {
        "gray7": "gray7",
      },
      "current": "dark",
    }
  }
>
  <ForwardRef(render)>
    <div
      className="emotion-0 emotion-1"
    />
  </ForwardRef(render)>
</Slot>
`;

exports[`renders the expected markup 1`] = `
.emotion-0 {
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.86em;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background-color: #FFFFFF;
}

<Footer>
  <nav
    className="emotion-0 emotion-1"
  >
    Lorem ipsum
  </nav>
</Footer>
`;

exports[`renders the expected markup basic Footer.Slot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<Slot>
  <ForwardRef(render)>
    <div
      className="emotion-0 emotion-1"
    />
  </ForwardRef(render)>
</Slot>
`;

exports[`renders the expected markup, transparent Footer 1`] = `
.emotion-0 {
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.86em;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background-color: gray7;
}

<Footer
  theme={
    Object {
      "colors": Object {
        "gray7": "gray7",
      },
      "current": "dark",
    }
  }
>
  <nav
    className="emotion-0 emotion-1"
  >
    Lorem ipsum
  </nav>
</Footer>
`;

exports[`should apply extra padding to first and last Slot child elements 1`] = `
.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  margin-left: 5px;
  margin-right: 5px;
}

.emotion-0:first-of-type {
  margin-left: calc(5px * 2);
}

.emotion-0:last-of-type {
  margin-right: calc(5px * 2);
}

<Slot>
  <ForwardRef(render)>
    <div
      className="emotion-5 emotion-6"
    >
      <button
        className="emotion-0"
      >
        Button 1
      </button>
      <button
        className="emotion-0"
      >
        Button 2
      </button>
      <button
        className="emotion-0"
      >
        Button 3
      </button>
      <button
        className="emotion-0"
      >
        Button 4
      </button>
      <button
        className="emotion-0"
      >
        Button 5
      </button>
    </div>
  </ForwardRef(render)>
</Slot>
`;
